cmd/opampsupervisor/e2e_test.go

```diff
@@ -0,0 +1,319 @@
+	onConnectedFunc := callbacks.OnConnectedFunc
+	callbacks.OnConnectedFunc = func(conn types.Connection) {
+		agentConn.Store(conn)
+		isAgentConnected.Store(true)
+		connectedChan <- true
+		if onConnectedFunc != nil {
+			onConnectedFunc(conn)
+		}
+	}
+	onConnectionCloseFunc := callbacks.OnConnectionCloseFunc
+	callbacks.OnConnectionCloseFunc = func(conn types.Connection) {
+		isAgentConnected.Store(false)
+		connectedChan <- false
+		if onConnectionCloseFunc != nil {
+			onConnectionCloseFunc(conn)
+		}
+	}
+	handler, _, err := s.Attach(server.Settings{
+		Callbacks: server.CallbacksStruct{
+			OnConnectingFunc: connectingCallback(callbacks),
+		},
+	})
+	require.NoError(t, err)
+	mux := http.NewServeMux()
+	mux.HandleFunc("/v1/opamp", handler)
+	httpSrv := httptest.NewServer(mux)
+
+	shutdown := func() {
+		t.Log("Shutting down")
+		err := s.Stop(context.Background())
+		assert.NoError(t, err)
+		httpSrv.Close()
+	}
+	send := func(msg *protobufs.ServerToAgent) {
+		if !isAgentConnected.Load() {
+			require.Fail(t, "Agent connection has not been established")
+		}
+
+		agentConn.Load().(types.Connection).Send(context.Background(), msg)
+	}
+	t.Cleanup(func() {
+		waitForConnection(connectedChan, false)
+		shutdown()
+	})
+	return httpSrv.Listener.Addr().String(), connectedChan, send
+}
+
+func newSupervisor(t *testing.T, configType string, extraConfigData map[string]string) *supervisor.Supervisor {
+	wd, err := os.Getwd()
+	require.NoError(t, err)
+	tpl, err := os.ReadFile(path.Join(wd, "testdata", "supervisor", "supervisor_"+configType+".yaml"))
+	require.NoError(t, err)
+
+	templ, err := template.New("").Parse(string(tpl))
+	require.NoError(t, err)
+
+	var buf bytes.Buffer
+	var extension string
+	if runtime.GOOS == "windows" {
+		extension = ".exe"
+	}
+	configData := map[string]string{
+		"goos":      runtime.GOOS,
+		"goarch":    runtime.GOARCH,
+		"extension": extension,
+	}
+
+	for key, val := range extraConfigData {
+		configData[key] = val
+	}
+	err = templ.Execute(&buf, configData)
+	require.NoError(t, err)
+	cfgFile, _ := os.CreateTemp(t.TempDir(), "config_*.yaml")
+	_, err = cfgFile.Write(buf.Bytes())
+	require.NoError(t, err)
+
+	s, err := supervisor.NewSupervisor(zap.NewNop(), cfgFile.Name())
+	require.NoError(t, err)
+
+	return s
+}
+
+func TestSupervisorStartsCollectorWithRemoteConfig(t *testing.T) {
+	var agentConfig atomic.Value
+	addr, connected, send := newServer(
+		t,
+		defaultConnectingHandler,
+		server.ConnectionCallbacksStruct{
+			OnMessageFunc: func(_ types.Connection, message *protobufs.AgentToServer) *protobufs.ServerToAgent {
+				if message.EffectiveConfig != nil {
+					config := message.EffectiveConfig.ConfigMap.ConfigMap[""]
+					if config != nil {
+						agentConfig.Store(string(config.Body))
+					}
+				}
+
+				return &protobufs.ServerToAgent{}
+			},
+		})
+
+	s := newSupervisor(t, "basic", map[string]string{"url": addr})
+	defer s.Shutdown()
+
+	waitForConnection(connected, true)
```

Comment:
Does this wait for Supervisor to connect to OpAMP server? If yes it would be nice to say it in a comment.

Reply:
I've updated the function name here and put a comment on the function to document it.